Zephyr cmake module tools cleanup

tejlmand opened the pull request:
Followup on: #41301

This is a rework on the tools handling in Zephyr CMake build system.

The main purpose is to achieve a better differentiation between tools / compiler lookup and general build system infrastructure.

This allows for easier re-usability of tools lookup and selection.
It further paves the way for allowing improved sharing of build settings when using sysbuild.

tejlmand committed:
cmake: cleanup search for devicetree compiler, dtc

Cleanup CMake code by moving devicetree compiler detection to dedicated
FindDtc.cmake module.

This allows us to re-use existing CMake find_package() functionality in
module mode and thus reduce CMake logic.

It further has the benefit of being able to directly specify dtc version
like this:
> find_package(Dtc 1.4.6)

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: move testing of compile flag in environment to top CMakeLists.txt

This moves the test of compile flags set in environment from generic
toolchain to top-level ${ZEPHYR_BASE}/CMakeLists.txt.

In order to cleanup and make Zephyr CMake modules more generic then
the testing of specific compile flags and warn user is not really a
toolchain module responsible but a implementation detail of the project
itself as to how compile flags should be applied in a project.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: rework of host tools and generic toolchain handling

Follow-up: #41301

This commit is a rework and cleanup of the tools handling in Zephyr
CMake build system.

Instead of directly loading code a CMake modules for tool lookup, the
host tools now follows the CMake `find_package()` pattern for finding
programs / tools in module mode.

This makes it more clear which modules are responsible for finding tools
and which modules provides build integration / features.

The following tools can now be found using `find_package()`:
- Zephyr-sdk        : find_package(Zephyr-sdk <version>)
- Generic host tools: find_package(HostTools)

This further allows us to decouple the `verify-toolchain` CMake script
part required by `twister` into a tool lookup module and a dedicated
CMake script which utilizes the lookup module.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: rework of target toolchain handling

Follow-up: #41301

This commit is a rework and cleanup of the target toolchain handling in
Zephyr CMake build system.

Instead of directly loading code a CMake modules for target toolchain
lookup, the target toolchain now follows the CMake `find_package()`
pattern for finding the target toolchain.

This makes it more clear which modules are responsible for finding tools
and which modules provides build integration / features.

The following tools can now be found using `find_package()`:
- Target toolchain: find_package(TargetTools)

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm commented on cmake/modules/FindHOST_TOOLS.cmake:
```diff
@@ -0,0 +1,124 @@
+# FindHOST_TOOLS module for locating a set of tools to use on the host for
+# Zephyr development.
+#
+# This module will lookup following tools for Zephyr development:
```
```suggestion
# This module will lookup the following tools for Zephyr development:
```

tejlmand replied:
Fixed

nordicjm commented on cmake/modules/FindHOST_TOOLS.cmake:
```diff
@@ -0,0 +1,124 @@
+# Set to 'IMGTOOL-NOTFOUND' if imgtool was not found.
+#
+# 'HOST_TOOLS_FOUND'
+#  True if all required host tools was found.
```
```suggestion
# True if all required host tools was found.
```

tejlmand replied:
Fixed

nordicjm commented on cmake/modules/FindHOST_TOOLS.cmake:
```diff
@@ -0,0 +1,124 @@
+    set(ZEPHYR_TOOLCHAIN_VARIANT cross-compile)
+endif()
+
+find_package(Zephyr-sdk 0.15)
```
Move to a variable near the top like in the original e.g.:
```
# This is the minimum required Zephyr SDK version.
set(TOOLCHAIN_ZEPHYR_MINIMUM_REQUIRED_VERSION 0.15)
```

tejlmand replied:
actually this was done on purpose.

Originally the reason to use a variable was because the minimum version was used in two places, however the disadvantage of the variable is that you must know its name to find it or do a two step search.

Whereas, with `find_package(Zephyr-sdk <version>)` you can search directly for `Zephyr-sdk` which is the package name, and then update the required version directly there.
Also much more readable when reading code, cause you can read the version immediately.

As bonus info the intention is to deprecate and remove those lines, so toolchain configuration gets aligned:
https://github.com/zephyrproject-rtos/zephyr/blob/afe878b9c7868757e1418ac7584b559bcbda3615/cmake/modules/FindHOST_TOOLS.cmake#L71-L81

and to see if this can be cleaned up as well:
https://github.com/zephyrproject-rtos/zephyr/blob/afe878b9c7868757e1418ac7584b559bcbda3615/cmake/modules/FindHOST_TOOLS.cmake#L64-L69

That will move the `find_package(Zephyr-sdk 0.15) further up, and everything becomes even more readable.
But it takes two Zephyr releases before deprecated code can be removed.

nordicjm commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -103,10 +87,10 @@ if(NOT DEFINED ZEPHYR_TOOLCHAIN_VARIANT)
     #  Note: When CMake minimum version becomes >= 3.17, change this loop into:
     #    foreach(version config IN ZIP_LISTS Zephyr-sdk_CONSIDERED_VERSIONS Zephyr-sdk_CONSIDERED_CONFIGS)
     set(error_msg "The Zephyr SDK version you are using is not supported, please update your SDK.\n")
-    set(missing_version "You need SDK version ${TOOLCHAIN_ZEPHYR_MINIMUM_REQUIRED_VERSION} or newer.")
```
When I force 0.14.2 I don't see this, I get this output:
```
CMake Error at /tmp/aa/zephyr/cmake/modules/FindZephyr-sdk.cmake:61 (find_package):
  Could not find a configuration file for package "Zephyr-sdk" that is
  compatible with requested version "0.15".

  The following configuration files were considered but not accepted:

    /opt/zephyr-sdk-0.14.2/cmake/Zephyr-sdkConfig.cmake, version: 0.14.2
```

tejlmand replied:
actually the code flow is not changed with this PR.
And when using CMake package, we rely on the CMake message printing, not this code snippet.

This is code which originates all the way back to when Zephyr SDK 0.11.x was still supported:
https://github.com/zephyrproject-rtos/zephyr/blob/7f970093b33cf794352324ce820d82ea575bec0c/cmake/toolchain/zephyr/host-tools.cmake#L76

I believe this code can actually be removed now, but will examine closer to make sure there are no pifalls.

Thanks for spotting.
When using the compiler from Zephyr SDK, that is: `ZEPHYR_TOOLCHAIN_VARIANT==zephyr|<none>` then the standard messages from `find_package()` will be printed.

But if you are combining a 3rd party toolchain with host tools from Zephyr SDK, then this code gets activated, for example:

tejlmand replied:
code block removed.

Also with this change https://github.com/zephyrproject-rtos/zephyr/commit/07e4d085544c53ebd5d2c44101d40e0426389db1 this code block can no longer be reached.

nordicjm reviewed (commented):
Filenames have gone from snake case to camel case

tejlmand commented:
> Filenames have gone from snake case to camel case

@nordicjm and i'm very much in doubt what to use.

My preference is lower case, like we do for other Zephyr CMake modules, `extensions.cmake`, `dts.cmake`, `kconfig.cmake`.

But CMake request `Find<modulename>.cmake` which then gives `FindDTC.cmake` vs `Finddtc.cmake` or `FindHOST_TOOLS.cmake` vs. `Findhost_tools.cmake` vs. `FindHostTools.cmake`.

The package name must the be used as:
`find_package(DTC)` vs `find_package(dtc)`
`find_package(HOST_TOOLS)` vs `find_package(host_tools)` vs `find_package(HostTools)`

we include other modules as:
https://github.com/zephyrproject-rtos/zephyr/blob/ac7e679ae503af1fd355e67db3f5e0adfb7ad795/cmake/modules/dts.cmake#L5-L8

so from that point of view, I would prefer `dtc` and `host_tools`.

but now the fun begins.
CMake requires a packages to set the variable `<packagename>_FOUND` to true.

This means that in code we would then need to test like this:
```
if(dtc_FOUND)
...

if(host_tools_FOUND)
...
```

but I consider those variables global, meaning in CMake code I would prefer:
```
if(DTC_FOUND)

if(HOST_TOOLS_FOUND)
```

I believe most users might see or need to test for `if(DTC_FOUND)` and similar, so therefore I decided to use the upper case in filename to get clean code in CMake.

But there is no good choice here, only bad and worse.
And which is the bad and which is the worse, I can't tell.

Open to change naming, but at least this is the explanation.

Ref: https://cmake.org/cmake/help/latest/command/find_package.html#basic-signature
> Regardless of the mode used, a <PackageName>_FOUND variable will be set to indicate whether the package was found.

tejlmand committed:
cmake: remove unreachable Zephyr-SDK lookup code

Removing Zephyr-SDK code for searching and printing messages when Zephyr
SDK is not found in the system.

After Zephyr SDK has been made available on all platforms, including
MacOS and Windows, then `find_package(Zephyr-sdk)` is always invoked
and thus either `find_package(Zephyr-sdk)` will fail or
ZEPHYR_TOOLCHAIN_VARIANT will be defined.

Meaning the removed code block in this commit can no longer be reached.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

stephanosio commented:
Rebased to fix CI failure

tejlmand commented:
> But there is no good choice here, only bad and worse.
> And which is the bad and which is the worse, I can't tell.

@nordicjm Did some more thinking on this.
If agreeing, then I will change into how CMake does this for `FindGit.cmake` and `FindUnixCommands.cmake`, which will result in:
`FindDtc.cmake` and `FindHostTools.cmake`.

The `FindGit.cmake` defines `GIT_FOUND` is addition to the required `Git_FOUND`, which means users can use either, and following this will allow for `Dtc_FOUND` and `DTC_FOUND` as well as `HostTools_FOUND` and `HOSTTOOLS_FOUND`.
This is also common behavior for handling of package standard arg for compatibility reasons.

Ref: https://cmake.org/cmake/help/latest/module/FindGit.html
> `Git_FOUND`, `GIT_FOUND`
>    True if the Git command-line client was found.

tejlmand commented:
> If agreeing, then I will change into how CMake does this for `FindGit.cmake` and `FindUnixCommands.cmake`, which will result in:
> `FindDtc.cmake` and `FindHostTools.cmake`.

Updated to be closer aligned with general CMake approach, and supporting both `Dtc_FOUND` and `DTC_FOUND`, same for `HostTools_FOUND` and `HOSTTOOLS_FOUND`.

stephanosio commented on cmake/modules/dts.cmake:
```diff
@@ -5,7 +5,7 @@ include_guard(GLOBAL)
 include(python)
 include(boards)
-include(generic_toolchain)
+find_package(HostTools)
```
any reason we are doing `find_package(HostTools)` in `dts.cmake`?

Also why are we handling the host tools as one big package? Should there not be individual packages for: gperf, openocd, bossac, ...

tejlmand replied:
because `dts.cmake` requires a generic toolchain (C compiler) which is provided by host tools.

> Also why are we handling the host tools as one big package? 

I did consider if it should be split, but that would result in one of following:
1. Each place someone needs one of the tools, they must now call `FindPackage(<tool>)` themselves first, for example in those places:
    https://github.com/zephyrproject-rtos/zephyr/blob/0cd311c4a1ebad126bdba7858a279c2bb00f7b04/CMakeLists.txt#L1013
    https://github.com/zephyrproject-rtos/zephyr/blob/0cd311c4a1ebad126bdba7858a279c2bb00f7b04/cmake/flash/CMakeLists.txt#L58
2. Create a single module which load all needed and available tools in a single location as general builds system configuration.

Going for 1. would soon become messy, whereas 2. is the solution we have today and this PR just tries to cleanup and places more complex handling like `DTC` detection in dedicated module.

Having 2. ensures that host tools needed for Zephyr development can be loaded using a single code line seen from the caller, not having to consider when a new tool is used.

I will be looking into the generic compiler later, so that hopefully `FindHostTools.cmake` purely becomes something like this in future:
```
find_program(<programA> <a>) # For simple program lookup
find_program(<programB> <b>)

find_packagage(<programC> <version>) # More complex lookup where version and/or other complex handling is needed.
```

It also follows the principle of finding a collection of tools, like: `find_package(UnixCommand)`
Ref: https://cmake.org/cmake/help/latest/module/FindUnixCommands.html

stephanosio commented on cmake/modules/FindHostTools.cmake:
```diff
@@ -0,0 +1,125 @@
+    set(ZEPHYR_TOOLCHAIN_VARIANT cross-compile)
+endif()
+
+find_package(Zephyr-sdk 0.15)
```
In CamelCase, `ZephyrSdk` would be a more appropriate name.

tejlmand replied:
yes, but I would like to have this file aligned with the Zephyr-sdk package from here:
https://github.com/zephyrproject-rtos/sdk-ng/blob/main/cmake/Zephyr-sdkConfig.cmake

so if we adjust `Zephyr-sdk` to `ZephyrSdk` then I will happily change.

nordicjm commented on cmake/modules/FindHostTools.cmake:
```diff
@@ -0,0 +1,125 @@
+    set(ZEPHYR_TOOLCHAIN_VARIANT cross-compile)
+endif()
+
+find_package(Zephyr-sdk 0.15)
```
Would still prefer this be near the top of the file so it's obvious if it's opened in a text editor that it can be changed vs half way down the file after non-configurable parts

tejlmand replied:
still not convinced, but I introduced a `Deprecated` module, which allows us to move deprecated code to a centralized place, and then load it where needed until we completely remove it.
See commits: 83cfbd6445ef34fc278117e1107c5dd7eebd6f54 and 9a187e2ec842891b59dbfc8accaedb3696c5dbf5

Deprecated code can now be loaded as:
```
find_package(Deprecated COMPONENT <deprecated_name>)
```
So now code looks much cleaner:
https://github.com/zephyrproject-rtos/zephyr/blob/9a187e2ec842891b59dbfc8accaedb3696c5dbf5/cmake/modules/FindHostTools.cmake#L52-L66

using a deprecated module here also has the benefit that users / CI / higher level config can completely disable deprecated code as if it was never there.
For example:
```
cmake ... -DCMAKE_DISABLE_FIND_PACKAGE_Deprecated=1
```
will completely disable all deprecated code, just as if it had been removed already.

Ref: https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html

nordicjm commented on cmake/modules/FindTargetTools.cmake:
```diff
@@ -1,6 +1,31 @@
+# Set to 'CMAKE_C_COMPILER-NOTFOUND' if no C compiler was found.
+#
+# 'TargetTools_FOUND', 'TARGETTOOLS_FOUND'
+#  True if all required host tools was found.
```
```suggestion
# True if all required host tools were found.
```

nordicjm commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -0,0 +1,90 @@
+# Zephyr toolchain variant to use if not defined already.
+#
+# 'Zephyr-sdk_FOUND'
+#  True if the Zephyr SDK was found.
```
```suggestion
# True if the Zephyr SDK was found.
```

tejlmand committed:
cmake: introduce Deprecated module

The deprecated CMake module allows Zephyr to move deprecated code to a
centralized location from where it's both easy to get an overview of
deprecated code, and remove the code when it has been deprecated long
enough.

It further improve existing code, as the real CMake code now can be
changed from n-lines of deprecated code into a single line of the form:
find_package(Deprecated COMPONENTS <deprecated_name>)

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: move deprecated code to FindDeprecated.cmake

Move deprecated XCC_USE_CLANG setting to deprecated module.

Deprecated and move the use of CROSS_COMPILE to deprecated module as we
have had ZEPHYR_TOOLCHAIN_VARIANT==cross-compile for a long time.

Deprecated code is now loaded with a single line:
> find_package(Deprecated COMPONENTS XCC_USE_CLANG CROSS_COMPILE)

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

nordicjm commented on CMakeLists.txt:
```diff
@@ -131,6 +131,17 @@ foreach(optional_include_dir
+  if(DEFINED ENV{${var}})
+    message(WARNING "The environment variable '${var}' was set to $ENV{${var}}, "
+                    "but Zephyr ignores flags from the environment. Use 'cmake "
+		    "-DEXTRA_${var}=$ENV{${var}}' instead."
```
Alignment here is off

tejlmand replied:
note to self.
Check formatting when copying other peoples code to new location 😆 

Fixed.

nordicjm commented on cmake/modules/FindZephyr-sdk.cmake:
```diff
@@ -0,0 +1,90 @@
+    endif()
+  else()
+    find_package(Zephyr-sdk ${Zephyr-sdk_FIND_VERSION} REQUIRED QUIET CONFIG PATHS
+                 /usr
+		 /usr/local
+		 /opt
+		 $ENV{HOME}
+		 $ENV{HOME}/.local
+		 $ENV{HOME}/.local/opt
+		 $ENV{HOME}/bin)
```
Not sure if it's the github viewer but these look to have different alignments
